tensorflow/compiler/mlir/xla/transforms/hlo_outline_fusion.cc

```diff
@@ -0,0 +1,557 @@
+  // larger combined cluster is represented by `cluster_from`'s ID in
+  // `cycles_graph_`.
+  bool MergeClusters(Cluster* cluster_from, Cluster* cluster_to) {
+    int from = cluster_from->cycles_graph_node_id();
+    int to = cluster_to->cycles_graph_node_id();
+
+    auto optional_merged_node = cycle_detector_.ContractEdge(from, to);
+    if (!optional_merged_node.has_value()) {
+      llvm::dbgs() << "Could not contract " << from << " -> " << to
+                   << " because contracting the edge would create a cycle.";
+      return false;
+    }
+
+    // Merge the clusters.
+    cluster_from->Merge(cluster_to);
+    cluster_from->set_cycles_graph_node_id(optional_merged_node.value());
+
+    // Merge the UnionFind<Cluster*>.
+    cluster_for_node_[from].Merge(&cluster_for_node_[to]);
+
+    return true;
+  }
+
+  template <typename FnTy>
+  bool ForEachEdgeInPostOrder(FnTy fn) {
+    bool changed = false;
+    for (int32 node : cycle_detector_.AllNodesInPostOrder()) {
+      Cluster* cluster_from = GetClusterForCyclesGraphNode(node);
+      if (!cluster_from) {
+        continue;
+      }
+
+      // Make a copy of the set of successors because we may modify the graph in
+      // TryToContractEdge.
+      std::vector<int32> successors_copy =
+          cycle_detector_.SuccessorsCopy(cluster_from->cycles_graph_node_id());
+
+      for (int to : successors_copy) {
+        Cluster* cluster_to = GetClusterForCyclesGraphNode(to);
+        if (!cluster_to) {
+          continue;
+        }
+
+        bool contracted_edge = fn(cluster_from, cluster_to);
+        changed |= contracted_edge;
+      }
+    }
+
+    return changed;
+  }
+
+  SmallVector<Value, 4> GetResultsOfFusedPattern(Cluster* from, Cluster* to) {
+    auto fused_pattern = MergeFusionPattern(
+        from->fused_pattern(), to->fused_pattern());
+    return GetOutputsOfFusionPattern(fused_pattern);
+  }
+
+  bool TryToContractEdge(Cluster* from, Cluster* to) {
+    int node_to = to->cycles_graph_node_id();
+    int node_from = from->cycles_graph_node_id();
+
+    // Both node_to and node_from should be fusible
+    if (!IsFusible(op_list_[node_to]) || !IsFusible(op_list_[node_from])) {
+      return false;
+    }
+
+    // ReduceOp is fusible only when it's the output of the fusion pattern.
+    if (isa<xla_hlo::ReduceOp>(op_list_[node_from])) {
+      return false;
+    }
+
+    // All outputs of a kLoop fusion should be same.
+    auto results = GetResultsOfFusedPattern(from, to);
+    if (results.empty()) return false;
+
+    auto get_workload_shape = [] (Value v) {
+      if (isa<xla_hlo::ReduceOp>(v.getDefiningOp())) {
+        v = v.getDefiningOp()->getOperand(0);
+      }
+      return v;
+    };
+
+    Value ref = get_workload_shape(results[0]);
+    if (!llvm::all_of(results, [&] (Value result) {
+      Value val = get_workload_shape(result);
+      return shape_analysis_.HasSameShape(ref, val);
```

Comment:
You will have to hit constantly the hash map for `ref` on each iteration because of the way HasSameShape API is designed (I don't have a great immediate suggestion but this may be interesting to look at the encapsulation here: the abstraction induces extra cost)